compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp

```diff
@@ -691,12 +692,156 @@ FailureOr<Flow::WorkloadBuilder> getWorkloadBuilder(OpBuilder &builder,
+/// with "reduction" types.
+static SmallVector<ReassociationIndices> getCollapsibleLoops(
+    linalg::GenericOp genericOp) {
+  SmallVector<ReassociationIndices> contiguousLoops;
+
+  SmallVector<unsigned> pDims;
+  genericOp.getParallelDims(pDims);
+  if (pDims.size() < 2) return contiguousLoops;
+
+  llvm::SmallDenseSet<unsigned> pLoops(pDims.begin(), pDims.end());
+
+  auto hasAllMapsSameSequence = [&](AffineExpr preExpr, AffineExpr nextExpr) {
+    for (AffineMap map : genericOp.getIndexingMapsArray()) {
+      bool foundSeq = false;
+      for (auto [index, resultExpr] : llvm::enumerate(map.getResults())) {
+        if (resultExpr == nextExpr) {
+          foundSeq = (index > 0 && preExpr == map.getResult(index - 1));
+          break;
+        }
+      }
+      if (!foundSeq) return false;
+    }
+    return true;
+  };
+
+  ReassociationIndices range;
+  AffineExpr preExpr;
+  for (auto nextExpr : genericOp.getIndexingMapsArray().front().getResults()) {
+    unsigned pos = nextExpr.cast<AffineDimExpr>().getPosition();
+    if (!range.empty()) {
+      if (!hasAllMapsSameSequence(preExpr, nextExpr) || !pLoops.count(pos)) {
+        if (range.size() > 1)
+          contiguousLoops.push_back({range.begin(), range.end()});
+        range.clear();
+      }
+    }
+    preExpr = nextExpr;
+    if (pLoops.count(pos)) range.push_back(pos);
+  }
+  if (range.size() > 1) contiguousLoops.push_back(range);
+
+  // Couldn't find anything to collapse
+  if (contiguousLoops.empty()) return contiguousLoops;
+  LLVM_DEBUG({
+    llvm::dbgs() << "Collapsing parallel loops: ";
+    for (auto indices : contiguousLoops) {
+      llvm::dbgs() << "[";
+      for (auto idx : indices) llvm::dbgs() << idx << ",";
+      llvm::dbgs() << "]\t";
+    }
+    llvm::dbgs() << "\n";
+  });
+
+  return contiguousLoops;
+}
+
+/// Collapse possible dimension of the given linalg.generic and return the
+/// new one
+static FailureOr<linalg::GenericOp> collapseLinalgGeneric(
+    TensorDimTrackingRewriter &rewriter, linalg::GenericOp genericOp) {
+  SmallVector<ReassociationIndices> collapseIndices =
+      getCollapsibleLoops(genericOp);
+
+  if (collapseIndices.empty()) return genericOp;
+
+  rewriter.setInsertionPoint(genericOp);
+  FailureOr<SmallVector<Value>> replacements =
+      mlir::linalg::collapseGenericOpIterationDims(genericOp, collapseIndices,
+                                                   rewriter);
+  if (failed(replacements) || replacements->empty()) {
+    return rewriter.notifyMatchFailure(genericOp,
+                                       "failed to collapse dimensions");
+  }
+
+  // Find and return collapsed linalg.generic
+  auto expandshapeOp =
+      replacements->front().getDefiningOp<tensor::ExpandShapeOp>();
+  if (!expandshapeOp) return failure();
+
+  auto newGenericOp =
+      expandshapeOp.getOperand().getDefiningOp<linalg::GenericOp>();
+  if (!newGenericOp) return failure();
```

Comment:
Can we make `collapseGenericOpIterationDims` return the new generic op and replacements? Maybe we can define a struct for it. (you don't have to fix it with the PR, we can fix it afterwards)

btw, should we look into all the replacements? Otherwise it may not work for multi-result cases?

Reply:
Good point, this will clean up the code. 
I should implement this first at upstream. I'll do this later.